File: haskell/glue/src/Glue/Lib/Math/Arithmetic/Div.hs
module Glue.Lib.Math.Arithmetic.Div where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Division function
-- Mirrors Haskell Glue.Lib.Math.Arithmetic.Div.div exactly
div :: IR Eval
div = NativeFunc divImpl

-- Division function implementation
-- Mirrors Haskell Glue.Lib.Math.Arithmetic.Div.divImpl exactly
divImpl :: IR Eval -> Eval (IR Eval)
divImpl left = pure $ NativeFunc (divBy left)

divBy :: IR Eval -> IR Eval -> Eval (IR Eval)
divBy left right = case (left, right) of
    (Integer a, Integer b) -> pure $ Float (fromIntegral a / fromIntegral b)
    (Integer a, Float b) -> pure $ Float (fromIntegral a / b)
    (Float a, Integer b) -> pure $ Float (a / fromIntegral b)
    (Float a, Float b) -> pure $ Float (a / b)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/arithmetic/div.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Division function
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Div.div exactly
final Ir div = IrNativeFunc(divImpl);

/// Division function implementation
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Div.divImpl exactly
Eval<Ir> divImpl(Ir left) {
  return Eval.pure(IrNativeFunc(divBy(left)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Div.divBy exactly
Eval<Ir> Function(Ir) divBy(Ir left) {
  return (Ir right) {
    return sequenceAll([eval(left), eval(right)]).flatMap((values) {
      return switch (values) {
        [IrInteger(value: final a), IrInteger(value: final b)] => Eval.pure(
          IrFloat(a / b),
        ),
        [IrInteger(value: final a), IrFloat(value: final b)] => Eval.pure(
          IrFloat(a / b),
        ),
        [IrFloat(value: final a), IrInteger(value: final b)] => Eval.pure(
          IrFloat(a / b),
        ),
        [IrFloat(value: final a), IrFloat(value: final b)] => Eval.pure(
          IrFloat(a / b),
        ),
        _ => throwError(wrongArgumentType(['number'])),
      };
    });
  };
}
